httpbp/client_middlewares.go

```diff
@@ -0,0 +1,239 @@
+		if resp.StatusCode >= http.StatusInternalServerError {
+			err = resp.Body.Close()
+			if err != nil {
+				return nil, err
+			}
+			return nil, err
+		}
+		return resp, nil
+	}
+}
+
+type retries struct {
+	next    http.RoundTripper
+	retries int
+}
+
+// Retries provides a retry middleware.
+func Retries(config Config) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return &retries{
+			next:    next,
+			retries: config.Retries,
+		}
+	}
+}
+
+func (r *retries) RoundTrip(req *http.Request) (*http.Response, error) {
+	remaining := r.retries + 1
+	for remaining > 0 {
+		resp, err := r.next.RoundTrip(req)
+		if err != nil {
+			return nil, err
+		}
+		return resp, nil
+	}
+	return r.next.RoundTrip(req)
+}
+
+type loadBalancer struct {
+	next     http.RoundTripper
+	hosts    []string
+	offset   uint64
+	attempts int
+}
+
+// LoadBalancer implements a round-robin load balancer.
+func LoadBalancer(config LoadBalancerConfig) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return &loadBalancer{
+			next:  next,
+			hosts: config.Hosts,
+		}
+	}
+}
+
+func (lb *loadBalancer) RoundTrip(req *http.Request) (*http.Response, error) {
+	remaining := lb.attempts
+	for remaining > 0 {
+		i := atomic.AddUint64(&lb.offset, 1)
+		hostIndex := i % uint64(len(lb.hosts))
+		req.Host = lb.hosts[hostIndex]
+		req.URL.Host = lb.hosts[hostIndex]
+
+		resp, err := lb.next.RoundTrip(req)
+		if err == gobreaker.ErrOpenState || err == gobreaker.ErrTooManyRequests {
+			remaining--
+		} else {
+			return resp, err
+		}
+	}
+	return lb.next.RoundTrip(req)
+}
+
+type maxConcurrency struct {
+	next           http.RoundTripper
+	maxConcurrency int64
+	activeRequests int64
+}
+
+// MaxConcurrency is a middleware to ensure that there will be only a maximum
+// number of requests concurrently in-flight at any given time.
+func MaxConcurrency(config Config) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return &maxConcurrency{
+			next:           next,
+			maxConcurrency: int64(config.MaxConcurrency),
+		}
+	}
+}
+
+func (m *maxConcurrency) RoundTrip(req *http.Request) (*http.Response, error) {
+	attemptedRequests := atomic.AddInt64(&m.activeRequests, 1)
+	defer atomic.AddInt64(&m.activeRequests, -1)
+
+	if m.maxConcurrency > 0 && attemptedRequests > m.maxConcurrency {
+		return nil, ErrConcurrencyLimit
+	}
+	return m.next.RoundTrip(req)
+}
+
+type monitorClient struct {
+	next http.RoundTripper
+	slug string
+}
+
+// MonitorClient is a HTTP client middleware that wraps HTTP requests in a
+// client span.
+func MonitorClient(config Config) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return &monitorClient{
+			next: next,
+			slug: config.Slug,
+		}
+	}
+}
+
+func (m *monitorClient) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	span, ctx := opentracing.StartSpanFromContext(
+		req.Context(),
+		m.slug,
```

Comment:
I believe this should be `m.slug + ".request"` to match the baseplate.py implementation.